### PR TITLE
Support to transform custom resource clientset

### PR DESCRIPTION
**Pull request by iawia002**
Currently clientgofix will not transform the methods in generated custom resource clientset, this PR has added support for this, eg:

```bash
$ clientgofix -custom-clientset=github.com/xx/xxrepo/pkg/clientset/versioned/typed/ ./...
```

### CODE REVIEW

**Comment by iawia002** [comment]
Both unit tests and local tests have passed:

```bash
go test -v ./...
?   	sigs.k8s.io/clientgofix	[no test files]
=== RUN   TestFix
=== RUN   TestFix/mod_vendor_v0.17.4
=== RUN   TestFix/mod_vendor_v0.17.4/comments
=== RUN   TestFix/mod_vendor_v0.17.4/dynamic
=== RUN   TestFix/mod_vendor_v0.17.4/expansions
=== RUN   TestFix/mod_vendor_v0.17.4/generated
=== RUN   TestFix/mod_vendor_v0.17.4/imports
=== RUN   TestFix/mod_vendor_v0.17.4/request
=== RUN   TestFix/mod_vendor_v0.17.4/scale
=== RUN   TestFix/mod_vendor_v0.17.4/testpackages
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/comments
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/dynamic
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/expansions
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/generated
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/imports
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/request
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/scale
=== RUN   TestFix/mod_vendor_v0.18.0-rc.1/testpackages
=== RUN   TestFix/mod_novendor_v0.17.4
=== RUN   TestFix/mod_novendor_v0.17.4/comments
=== RUN   TestFix/mod_novendor_v0.17.4/dynamic
=== RUN   TestFix/mod_novendor_v0.17.4/expansions
=== RUN   TestFix/mod_novendor_v0.17.4/generated
=== RUN   TestFix/mod_novendor_v0.17.4/imports
=== RUN   TestFix/mod_novendor_v0.17.4/request
=== RUN   TestFix/mod_novendor_v0.17.4/scale
=== RUN   TestFix/mod_novendor_v0.17.4/testpackages
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/comments
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/dynamic
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/expansions
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/generated
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/imports
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/request
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/scale
=== RUN   TestFix/mod_novendor_v0.18.0-rc.1/testpackages
=== RUN   TestFix/gopath_vendor_v0.17.4
=== RUN   TestFix/gopath_vendor_v0.17.4/comments
=== RUN   TestFix/gopath_vendor_v0.17.4/dynamic
=== RUN   TestFix/gopath_vendor_v0.17.4/expansions
=== RUN   TestFix/gopath_vendor_v0.17.4/generated
=== RUN   TestFix/gopath_vendor_v0.17.4/imports
=== RUN   TestFix/gopath_vendor_v0.17.4/request
=== RUN   TestFix/gopath_vendor_v0.17.4/scale
=== RUN   TestFix/gopath_vendor_v0.17.4/testpackages
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/comments
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/dynamic
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/expansions
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/generated
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/imports
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/request
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/scale
=== RUN   TestFix/gopath_vendor_v0.18.0-rc.1/testpackages
=== RUN   TestFix/gopath_novendor_v0.17.4
=== RUN   TestFix/gopath_novendor_v0.17.4/comments
=== RUN   TestFix/gopath_novendor_v0.17.4/dynamic
=== RUN   TestFix/gopath_novendor_v0.17.4/expansions
=== RUN   TestFix/gopath_novendor_v0.17.4/generated
=== RUN   TestFix/gopath_novendor_v0.17.4/imports
=== RUN   TestFix/gopath_novendor_v0.17.4/request
=== RUN   TestFix/gopath_novendor_v0.17.4/scale
=== RUN   TestFix/gopath_novendor_v0.17.4/testpackages
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/comments
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/dynamic
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/expansions
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/generated
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/imports
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/request
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/scale
=== RUN   TestFix/gopath_novendor_v0.18.0-rc.1/testpackages
--- PASS: TestFix (115.41s)
    --- PASS: TestFix/mod_vendor_v0.17.4 (14.73s)
        --- PASS: TestFix/mod_vendor_v0.17.4/comments (2.34s)
        --- PASS: TestFix/mod_vendor_v0.17.4/dynamic (1.00s)
        --- PASS: TestFix/mod_vendor_v0.17.4/expansions (2.04s)
        --- PASS: TestFix/mod_vendor_v0.17.4/generated (1.76s)
        --- PASS: TestFix/mod_vendor_v0.17.4/imports (1.75s)
        --- PASS: TestFix/mod_vendor_v0.17.4/request (1.34s)
        --- PASS: TestFix/mod_vendor_v0.17.4/scale (2.41s)
        --- PASS: TestFix/mod_vendor_v0.17.4/testpackages (1.30s)
    --- PASS: TestFix/mod_vendor_v0.18.0-rc.1 (13.97s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/comments (1.87s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/dynamic (1.17s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/expansions (1.93s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/generated (2.23s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/imports (1.88s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/request (1.18s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/scale (2.17s)
        --- PASS: TestFix/mod_vendor_v0.18.0-rc.1/testpackages (0.97s)
    --- PASS: TestFix/mod_novendor_v0.17.4 (13.21s)
        --- PASS: TestFix/mod_novendor_v0.17.4/comments (2.09s)
        --- PASS: TestFix/mod_novendor_v0.17.4/dynamic (1.12s)
        --- PASS: TestFix/mod_novendor_v0.17.4/expansions (1.69s)
        --- PASS: TestFix/mod_novendor_v0.17.4/generated (1.88s)
        --- PASS: TestFix/mod_novendor_v0.17.4/imports (2.23s)
        --- PASS: TestFix/mod_novendor_v0.17.4/request (1.15s)
        --- PASS: TestFix/mod_novendor_v0.17.4/scale (1.94s)
        --- PASS: TestFix/mod_novendor_v0.17.4/testpackages (1.11s)
    --- PASS: TestFix/mod_novendor_v0.18.0-rc.1 (13.92s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/comments (2.17s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/dynamic (1.54s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/expansions (1.90s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/generated (2.01s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/imports (1.96s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/request (1.34s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/scale (1.84s)
        --- PASS: TestFix/mod_novendor_v0.18.0-rc.1/testpackages (1.15s)
    --- PASS: TestFix/gopath_vendor_v0.17.4 (14.84s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/comments (2.12s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/dynamic (1.32s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/expansions (1.90s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/generated (2.13s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/imports (1.83s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/request (1.44s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/scale (1.92s)
        --- PASS: TestFix/gopath_vendor_v0.17.4/testpackages (1.63s)
    --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1 (16.08s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/comments (2.30s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/dynamic (2.12s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/expansions (1.77s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/generated (1.94s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/imports (2.16s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/request (1.21s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/scale (2.30s)
        --- PASS: TestFix/gopath_vendor_v0.18.0-rc.1/testpackages (1.37s)
    --- PASS: TestFix/gopath_novendor_v0.17.4 (13.22s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/comments (1.77s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/dynamic (1.24s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/expansions (1.70s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/generated (1.95s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/imports (1.67s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/request (1.05s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/scale (1.86s)
        --- PASS: TestFix/gopath_novendor_v0.17.4/testpackages (1.27s)
    --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1 (15.44s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/comments (2.08s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/dynamic (1.01s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/expansions (2.02s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/generated (2.14s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/imports (2.25s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/request (1.17s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/scale (2.16s)
        --- PASS: TestFix/gopath_novendor_v0.18.0-rc.1/testpackages (1.70s)
PASS
ok  	sigs.k8s.io/clientgofix/pkg	115.484s
```

/assign @liggitt

File: pkg/fix.go [hunk 1]
```diff
@@ -92,6 +93,11 @@ func (o *FixOptions) Run() error {
 		}
 	}
 
+	var customPrefixes []string
+	if o.CustomClientset != "" {
+		customPrefixes = strings.Split(o.CustomClientset, ",")
```
Comment (liggitt) [inline pkg/fix.go]: some sanity checking would be good, e.g. ensure there are no empty values, ensure the values end in a `/`

File: pkg/functions.go [hunk 2]
```diff
@@ -46,356 +46,375 @@ var dynamicClientPrefixes = []string{
 	"k8s.io/client-go/dynamic.",
 	"k8s.io/client-go/metadata.",
 }
-var generatedClientPrefixes = []string{
-	"k8s.io/client-go/kubernetes/typed/",
-	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/",
-	"k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/",
+
+func genClientPrefixes(customPrefixes []string) []string {
+	prefixeMap := map[string]struct{}{
```
Comment (liggitt) [inline pkg/functions.go]: nit: `prefixMap`

**Review by liggitt — COMMENTED** [review]
Thanks for the contribution. It would be good to exercise the custom client path in the unit tests as well.

**Comment by iawia002** [comment]
Hi @liggitt thanks for the review, all comments addressed and tests for custom clientset (the generated clientset is copied from [sample-controller](https://github.com/kubernetes/sample-controller/tree/master/pkg/generated/clientset/versioned)) is added, PTAL.

File: pkg/fix.go (hunk 1 above)
Comment (iawia002) [reply pkg/fix.go]: Some basic checks have been added

File: pkg/functions.go (hunk 2 above)
Comment (iawia002) [reply pkg/functions.go]: Fixed

File: pkg/fix.go [hunk 3]
```diff
@@ -92,6 +93,18 @@ func (o *FixOptions) Run() error {
 		}
 	}
 
+	var customPrefixes []string
+	for _, s := range strings.Split(o.CustomClientset, ",") {
+		s = strings.TrimSpace(s)
+		if s == "" {
+			continue
+		}
+		if s[len(s)-1] != '/' {
```
Comment (liggitt) [inline pkg/fix.go]: nit: `!strings.HasSuffix(s, "/")` would be easier to read

File: pkg/functions.go [hunk 4]
```diff
@@ -46,356 +46,375 @@ var dynamicClientPrefixes = []string{
...
+		"k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/": {},
+	}
+	for _, p := range customPrefixes {
+		prefixMap[p] = struct{}{}
+	}
+	prefixes := make([]string, 0, len(prefixMap))
+	for p := range prefixMap {
+		prefixes = append(prefixes, p)
+	}
+	return prefixes
```
Comment (liggitt) [inline pkg/functions.go:67]: nit: `sort.Strings` this to get deterministic output and run order

**Review by liggitt — COMMENTED** [review]
thanks, two nits, then lgtm

**Comment by iawia002** [comment]
Thanks for the review, all comments addressed and commits have been squashed into one, PTAL.

**Pull request merged**